File: sdks/typescript/src/core/pagination.ts
/**
 * 游标分页模块
 */

import { BaseAPIClient } from "./client";

export interface CursorPaginatorOptions {
  cursorParam?: string;
  limitParam?: string;
  defaultLimit?: number;
  responseCursorField?: string;
  responseItemsField?: string;
}

export class CursorPaginator<T> {
  private client: BaseAPIClient;
  private path: string;
  private method: string;
  private params: Record<string, any>;
  private body?: any;

  private cursorParam: string;
  private limitParam: string;
  private defaultLimit: number;
  private responseCursorField: string;
  private responseItemsField: string;

  private _cursor: string | null = null;
  private _hasMore: boolean = true;
  private _buffer: T[] = [];

  constructor(
    client: BaseAPIClient,
    path: string,
    method: string = "GET",
    params: Record<string, any> = {},
    body?: any,
    options: CursorPaginatorOptions = {}
  ) {
    this.client = client;
    this.path = path;
    this.method = method;
    this.params = params;
    this.body = body;

    this.cursorParam = options.cursorParam ?? "cursor";
    this.limitParam = options.limitParam ?? "limit";
    this.defaultLimit = options.defaultLimit ?? 20;
    this.responseCursorField = options.responseCursorField ?? "next_cursor";
    this.responseItemsField = options.responseItemsField ?? "items";
  }

  /**
   * 获取下一页数据
   */
  async nextPage(limit?: number): Promise<T[]> {
    if (!this._hasMore) {
      return [];
    }

    const params = { ...this.params };
    params[this.limitParam] = limit ?? this.defaultLimit;

    if (this._cursor) {
      params[this.cursorParam] = this._cursor;
    }

    const response: any = await this.client._request(
      this.method,
      this.path,
      params,
      this.body
    );

    if (Array.isArray(response)) {
      this._hasMore = false;
      return response as T[];
    }

    const items = response?.[this.responseItemsField] ?? [];
    this._cursor = response?.[this.responseCursorField] ?? null;
    this._hasMore = this._cursor !== null;

    return items as T[];
  }

  /**
   * 获取所有数据
   */
  async all(limit?: number): Promise<T[]> {
    const allItems: T[] = [];
    while (this._hasMore) {
      const items = await this.nextPage(limit);
      allItems.push(...items);
    }
    return allItems;
  }

  /**
   * 是否有更多数据
   */
  get hasMore(): boolean {
    return this._hasMore;
  }

  /**
   * 当前游标
   */
  get cursor(): string | null {
    return this._cursor;
  }

  /**
   * 异步迭代器支持
   */
  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    while (this._hasMore) {
      const items = await this.nextPage();
      for (const item of items) {
        yield item;
      }
    }
  }
}

File: sdks/python/bolt_prediction_sdk/core/pagination.py
"""
游标分页模块
"""

from typing import Any, Dict, List, Optional, AsyncIterator
from loguru import logger


class CursorPaginator:
    """游标分页器

    支持同步迭代和异步迭代两种方式遍历所有数据。
    """

    def __init__(
        self,
        client: Any,
        path: str,
        method: str = "GET",
        params: Optional[Dict[str, Any]] = None,
        body: Optional[Any] = None,
        cursor_param: str = "cursor",
        limit_param: str = "limit",
        default_limit: int = 20,
        response_cursor_field: str = "next_cursor",
        response_items_field: str = "items",
    ):
        self.client = client
        self.path = path
        self.method = method
        self.params = params or {}
        self.body = body
        self.cursor_param = cursor_param
        self.limit_param = limit_param
        self.default_limit = default_limit
        self.response_cursor_field = response_cursor_field
        self.response_items_field = response_items_field

        self._cursor: Optional[str] = None
        self._has_more: bool = True
        self._buffer: List[Any] = []

    async def next_page(self, limit: Optional[int] = None) -> List[Any]:
        """
        获取下一页数据

        Args:
            limit: 每页数量

        Returns:
            本页数据列表
        """
        if not self._has_more:
            return []

        params = dict(self.params)
        params[self.limit_param] = limit or self.default_limit

        if self._cursor:
            params[self.cursor_param] = self._cursor

        request_kwargs = {
            "method": self.method,
            "path": self.path,
            "params": params,
        }

        if self.body is not None:
            request_kwargs["json"] = self.body

        response = await self.client._request(**request_kwargs)

        if isinstance(response, dict):
            items = response.get(self.response_items_field, [])
            self._cursor = response.get(self.response_cursor_field)
            self._has_more = self._cursor is not None
        else:
            items = response
            self._has_more = False

        return items

    async def all(self, limit: Optional[int] = None) -> List[Any]:
        """
        获取所有数据

        Args:
            limit: 每页数量

        Returns:
            所有数据的列表
        """
        all_items: List[Any] = []
        while self._has_more:
            items = await self.next_page(limit)
            all_items.extend(items)
        return all_items

    async def __aiter__(self) -> AsyncIterator[Any]:
        """异步迭代器支持"""
        while self._has_more:
            items = await self.next_page()
            for item in items:
                yield item
